<?xml version="1.0"?>
<robot xmlns:xacro="http://wiki.ros.org/xacro" name="drone_delivery">
  <xacro:property name="distance" value="0.167"/>
  <xacro:property name="mass" value="1"/>
  <material name="blue">
    <color rgba= "0 0 1 1"/>
  </material>
  <material name="grey">
    <color rgba="0.75 0.75 0.75 1"/>
  </material>
  <material name="WHITE_MTL">
    <color rgba="0 255 255 1.0"/> <!-- rviz green color -->
  </material>
  
  <link
    name="base_link">
    <inertial>
      <origin
        xyz="0 0 0.02"
        rpy="0 0 0" />
      <mass
        value="${mass}" />
      <inertia
        ixx="${0.0833333 * mass * (distance * distance + distance/4 * distance/4)}"
        ixy="0"
        ixz="0"
        iyy="${0.0833333 * mass * (distance * distance + distance/4 * distance/4)}"
        iyz="0"
        izz="${0.0833333 * mass * (distance * distance + distance * distance)}" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/base_link.STL" />
      </geometry>
      <material name="WHITE_MTL"/>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/base_link.STL" />
      </geometry>
    </collision>
  </link>
  <gazebo reference="base_link">
    <material>Gazebo/Black</material>
  </gazebo>
  <link
    name="prop_1">
    <inertial>
      <origin
        xyz="1.38777878078145E-17 1.6252017232217E-05 1.38777878078145E-17"
        rpy="0 0 0" />
      <mass
        value="0.1" />
      <inertia
        ixx="6.26992620308679E-07"
        ixy="-3.30120444584961E-24"
        ixz="7.98941079078638E-09"
        iyy="6.38900379274268E-07"
        iyz="1.24797735656839E-23"
        izz="1.55888758686964E-08" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_1.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.286274509803922 0.662745098039216 0.329411764705882 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_1.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="motor_1"
    type="continuous">
    <origin
      xyz="-0.084853 0.084853 0.0335"
      rpy="1.5708 0 1.59" />
    <parent
      link="base_link" />
    <child
      link="prop_1" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-1000000"
      upper="1000000"
      effort="1000000"
      velocity="1000000" />
  </joint>
  <link
    name="prop_2">
    <inertial>
      <origin
        xyz="0 1.6252017232217E-05 -2.77555756156289E-17"
        rpy="0 0 0" />
      <mass
        value="0.1" />
      <inertia
        ixx="6.26992620308679E-07"
        ixy="2.22863875678516E-24"
        ixz="7.98941079078498E-09"
        iyy="6.38900379274269E-07"
        iyz="-7.82816974953869E-25"
        izz="1.55888758686964E-08" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_2.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.286274509803922 0.662745098039216 0.329411764705882 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_2.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="motor_2"
    type="continuous">
    <origin
      xyz="0.084853 0.084853 0.0335"
      rpy="1.5708 0 -3.102" />
    <parent
      link="base_link" />
    <child
      link="prop_2" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-1000000"
      upper="1000000"
      effort="1000000"
      velocity="1000000" />
  </joint>
  <link
    name="prop_3">
    <inertial>
      <origin
        xyz="2.77555756156289E-17 0.00301625201723221 -1.38777878078145E-17"
        rpy="0 0 0" />
      <mass
        value="0.1" />
      <inertia
        ixx="6.26992620308679E-07"
        ixy="3.59808441353406E-24"
        ixz="7.98941079078556E-09"
        iyy="6.38900379274268E-07"
        iyz="7.60544173089602E-24"
        izz="1.55888758686964E-08" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_3.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.286274509803922 0.662745098039216 0.329411764705882 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_3.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="motor_3"
    type="continuous">
    <origin
      xyz="0.084853 -0.084853 0.0305"
      rpy="1.5708 0 -1.5516" />
    <parent
      link="base_link" />
    <child
      link="prop_3" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-1000000"
      upper="1000000"
      effort="1000000"
      velocity="1000000" />

  </joint>
  <link
    name="prop_4">
    <inertial>
      <origin
        xyz="2.77555756156289E-17 0.00301625201723221 0"
        rpy="0 0 0" />
      <mass
        value="0.1" />
      <inertia
        ixx="6.26992620308678E-07"
        ixy="7.16859571560985E-24"
        ixz="7.98941079078488E-09"
        iyy="6.38900379274267E-07"
        iyz="2.55140400374983E-23"
        izz="1.55888758686963E-08" />
    </inertial>
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_4.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.286274509803922 0.662745098039216 0.329411764705882 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://drone_delivery/meshes/prop_4.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="motor_4"
    type="continuous">
    <origin
      xyz="-0.084853 -0.084853 0.0305"
      rpy="1.5708 0 0.039629" />
    <parent
      link="base_link" />
    <child
      link="prop_4" />
    <axis
      xyz="0 1 0" />
    <limit
      lower="-1000000"
      upper="1000000"
      effort="1000000"
      velocity="1000000" />
   
  </joint>

<!--    
  <link name="imu_link">
    <inertial>
      <mass value="0.001"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.000001" iyz="0" izz="0.0001"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
      <material name="red"/>  
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size=".01 .01 .01"/>
      </geometry>
    </collision>
  </link>
  <joint name="imu_joint" type="fixed">
    <axis xyz="1 0 0"/>      
    <origin xyz="0 0 0.05"/>
    <parent link="base_link"/>
    <child link="imu_link"/>
  </joint> -->

  <gazebo reference="prop_1">
    <material>Gazebo/Green</material>
  </gazebo>
  <gazebo reference="prop_2">
    <material>Gazebo/Green</material>
  </gazebo>
  <gazebo reference="prop_3">
    <material>Gazebo/Green</material>
  </gazebo>
  <gazebo reference="prop_4">
    <material>Gazebo/Green</material>
  </gazebo>
  <gazebo reference="imu_link">
    <material>Gazebo/Orange</material>
  </gazebo>

  <!-- Transmission -->
  
  <ros2_control name="GazeboSystem" type="system">
  <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
  </hardware>

  <joint name="motor_1">
      <command_interface name="velocity">
          <param name="min">-1000.0</param>
          <param name="max">1000.0</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
  </joint>


  <joint name="motor_2">
      <command_interface name="velocity">
          <param name="min">-1000.0</param>
          <param name="max">1000.0</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
  </joint>

  <joint name="motor_3">
      <command_interface name="velocity">
          <param name="min">-1000.0</param>
          <param name="max">1000.0</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
  </joint>

  <joint name="motor_4">
      <command_interface name="velocity">
          <param name="min">-1000.0</param>
          <param name="max">1000.0</param>
      </command_interface>
      <state_interface name="position"/>
      <state_interface name="velocity"/>
  </joint>


  </ros2_control>

  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
          <parameters>$(find drone_delivery)/config/control.yaml</parameters>
    </plugin>
    <plugin name="drone_plugin" filename="libdrone.so">
    <!-- Name of the base/body link of the drone to compute odometry -->
      <drone_frame>base_link</drone_frame>
      <updateRate>100</updateRate>
      
      <!-- Keep this true if you want to visualize transforms for RViz -->
      <publishTf>true</publishTf>
      <!-- Adjust the coefficient according to need -->

      <rotorThrustCoeff>0.00025</rotorThrustCoeff>
      <rotorTorqueCoeff>0.0000074</rotorTorqueCoeff>
    </plugin>
    
  </gazebo>


</robot>
